substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GreyToBlackObjRefVisitor.java

```diff
@@ -140,6 +160,7 @@ public boolean visitObjectReferenceInline(final Pointer objRef, final int innerO
         } else {
             getCounters().noteUnmodifiedReference();
         }
+        HeapImpl.getHeapImpl().dirtyCardIfNecessary(holderObject, objRef, copy);
```

Comment:
Please add a comment that describes why it is crucial to dirty the card even though the reference was not updated (i.e., `copy == obj`).

Reply:
done, it doesn't need dirty the card when `copy == obj`, `dirtyCardIfNecessary ` should only called when the reference was updated

Reply:
Are you sure about that? What if an unaligned chunk is promoted? In that case, the object won't move, only the chunk will get assigned to a different space.

Reply:
Yes, you are right, the unaligned object promotion will not update the reference, I add a comment to describe this